kernel: Add C++ main() support

stephanosio opened the pull request:
This series adds the support for using C++ compatible main() declaration in the Zephyr kernel and allows Zephyr applications to define `int main(void)` in C++ source files as required by the C++ standard.

---

kernel: Add C++ main() support 

The C++ standard requires the main() function to have the return type
of 'int' and does not allow the main() to be defined with the 'void'
return type. Moreover, GCC goes as far as to emit a hard error when the
'::main()' has the return type of `void`.

This commit introduces an option to instruct the Zephyr kernel to call
the 'int main(void)' instead of the 'void main(void)' in case a Zephyr
application defines main() in a C++ source file.

p.s. `CONFIG_CPP_MAIN` is not `default y` when C++ support is enabled because it is common to have `main()` implemented in a C source file for Zephyr applications even when C++ support is enabled.

stephanosio committed:
kernel: Add C++ main() support

The C++ standard requires the main() function to have the return type
of 'int' and does not allow the main() to be defined with the 'void'
return type. Moreover, GCC goes as far as to emit a hard error when the
'::main()' has the return type of `void`.

This commit introduces an option to instruct the Zephyr kernel to call
the 'int main(void)' instead of the 'void main(void)' in case a Zephyr
application defines main() in a C++ source file.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio committed:
arch: posix: Declare _posix_zephyr_main with int return type

This commit updates the `_posix_zephyr_main` declaration to use the
return type of `int` instead of `void` when `CONFIG_CPP_MAIN=y` (i.e.
C++-compliant main() support is enabled) so that Zephyr applications
defining their main() in a C++ source file can make use of the proper
main() definition of `int main(void)` as required by the C++ standard.

Note that the forward declaration of `_posix_zephyr_main` is required
if and only if the main() is defined in a C++ source file (i.e. when
`CONFIG_CPP_MAIN=y`).

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio committed:
samples: cpp_synchronization: Enable CONFIG_CPP_MAIN

This commit updates the cpp_synchronization sample to enable
`CONFIG_CPP_MAIN` since it defines its main() in a C++ source file.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio committed:
samples: cpp_synchronization: Allow running on POSIX arch boards

This commit updates the cpp_synchronization to run on the "POSIX
architecture" boards such as `native_posix`, since the POSIX arch now
properly supports the main() defined in a C++ source file.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio committed:
samples: chre: Correct main() function prototype

C++ standard requires C++ main() to have the return type of 'int'.

This commit updates the CHRE sample to define main() as
`int main(void)` and enable  `CONFIG_CPP_MAIN`, which instructs the
Zephyr kernel to call the C++ main().

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio commented:
@yperess FYI, while working on [samples: chre: Correct main() function prototype](https://github.com/zephyrproject-rtos/zephyr/pull/51966/commits/25df2dec14f314e4cc2ee8fd637dc51475cbd1d3), I noticed that the CHRE sample is currently only being built for native_posix because of the malformed filter (`filter: not CONFIG_MINIMAL_LIBC`). Unless CHRE is supposed to be for native_posix only, the sample should be updated such that it (can) build for all boards (especially noting that there are no tests for CHRE at all).

Also even on native_posix, the sample reports a failure (it is not caught by twister because the regex in sample.yaml does not take this into account):

```
*** Booting Zephyr OS build zephyr-v3.2.0-1266-g25df2dec14f3 ***
Failed to set thread name to "CHRE": -38
Hello CHRE!
[00:00:00.000,000] <inf> chre: CHRE init, version: undefined
[00:00:00.000,000] <inf> chre: EventLoop start
[00:00:00.510,000] <dbg> chre: startNanoapp: Instance ID 1 assigned to app ID 0x0000000000000001
EchoApp::nanoappStart()
Starting EchoApp... SUCCESS
Nanoapp count=1
Finding instance ID... SUCCESS
Nanoapp count=1
Instance ID: 1
Sending event 1...
EchoApp::nanoappHandleEvent(sender_instance_id=0, event_type=1, event_data@(nil))
Event (1) complete!
[00:00:01.020,000] <err> chre: Refusing to unload system nanoapp
Ending EchoApp... FAIL
Goodbye!
EchoApp::nanoappEnd()
[00:00:01.020,000] <inf> chre: Exiting EventLoop
[00:00:01.020,000] <dbg> chre: deinit: CHRE deinit
```

yperess commented:
> @yperess FYI, while working on [samples: chre: Correct main() function prototype](https://github.com/zephyrproject-rtos/zephyr/pull/51966/commits/25df2dec14f314e4cc2ee8fd637dc51475cbd1d3), I noticed that the CHRE sample is currently only being built for native_posix because of the malformed filter (`filter: not CONFIG_MINIMAL_LIBC`). Unless CHRE is supposed to be for native_posix only, the sample should be updated such that it (can) build for all boards (especially noting that there are no tests for CHRE at all).
> 
> Also even on native_posix, the sample reports a failure (it is not caught by twister because the regex in sample.yaml does not take this into account):
> 
> ```
> *** Booting Zephyr OS build zephyr-v3.2.0-1266-g25df2dec14f3 ***
> Failed to set thread name to "CHRE": -38
> Hello CHRE!
> [00:00:00.000,000] <inf> chre: CHRE init, version: undefined
> [00:00:00.000,000] <inf> chre: EventLoop start
> [00:00:00.510,000] <dbg> chre: startNanoapp: Instance ID 1 assigned to app ID 0x0000000000000001
> EchoApp::nanoappStart()
> Starting EchoApp... SUCCESS
> Nanoapp count=1
> Finding instance ID... SUCCESS
> Nanoapp count=1
> Instance ID: 1
> Sending event 1...
> EchoApp::nanoappHandleEvent(sender_instance_id=0, event_type=1, event_data@(nil))
> Event (1) complete!
> [00:00:01.020,000] <err> chre: Refusing to unload system nanoapp
> Ending EchoApp... FAIL
> Goodbye!
> EchoApp::nanoappEnd()
> [00:00:01.020,000] <inf> chre: Exiting EventLoop
> [00:00:01.020,000] <dbg> chre: deinit: CHRE deinit
> ```

I'll update the filter. The "failure"is intended. You can't shut off a static nanoapp. I just wanted to verify that in the sample

stephanosio commented:
> I'll update the filter. The "failure"is intended. You can't shut off a static nanoapp. I just wanted to verify that in the sample

Thanks for clarifying. By the way, I tried enabling the sample for all boards and saw a build failure for Xtensa.

stephanosio committed:
doc: languages: cpp: Add notes about C++ main()

This commit adds notes about the C++ main() function prototype in the
C++ documentation.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

andyross reviewed (approved):
Seems harmless, though that's a lot of code just to match a function signature.
